apps/notifications/utils.py

```diff
@@ -0,0 +1,97 @@
+from django.core.mail import send_mail
+
+from apps.authentication.models import OnlineGroup as Group
+from apps.authentication.models import OnlineUser as User
+
+from .constants import PermissionType
+from .models import Notification, Permission, UserPermission
+from .tasks import dispatch_email_notification_task, dispatch_push_notification_task
+
+
+def send_message_to_users(
+    title: str,
+    content: str,
+    recipients: Iterable[User],
+    permission_type: PermissionType,
+    from_email=settings.DEFAULT_FROM_EMAIL,
+    image=None,
+    url="/",
+    tag=None,
+    force_override_dont_send_email=False,
+    force_override_dont_send_push=False,
+):
+    permission, created = Permission.objects.get_or_create(
+        permission_type=permission_type
+    )
+    for recipient in recipients:
+        # Make sure all permissions exists for the user before sending anything.
+        # Available permissions might have changed since the user last loaded their permission dashboard.
+        UserPermission.create_all_for_user(recipient)
+
+        user_permission = UserPermission.objects.get(
+            user=recipient, permission=permission
+        )
+
+        has_push_permission = permission.allow_push and (
+            (permission.force_push or user_permission.allow_push)
+            and not force_override_dont_send_push
+        )
+        has_email_permission = permission.allow_email and (
+            (permission.force_email or user_permission.allow_email)
+            and not force_override_dont_send_email
+        )
+
+        if not has_push_permission and not has_email_permission:
+            # Don't create notification at all if permission for none of the types is given
+            continue
+
+        notification = Notification.objects.create(
+            title=title,
+            body=content,
+            recipient=recipient,
+            from_email=from_email,
+            permission=permission,
+            image=image,
+            url=url,
+            tag=tag,
+        )
+
+        if has_push_permission:
+            dispatch_push_notification_task.delay(notification_id=notification.id)
+
+        if has_email_permission:
+            dispatch_email_notification_task.delay(notification_id=notification.id)
+
+
+def send_message_to_group(
+    title: str,
+    content: str,
+    group: Group,
+    from_email=settings.DEFAULT_FROM_EMAIL,
+    image=None,
+    url="/",
+    tag=None,
+):
+    send_mail(
+        subject=title,
+        message=content,
+        from_email=from_email,
+        recipient_list=[group.email],
+        fail_silently=False,
+    )
+
+    recipients = [member.user for member in group.members.all()]
+
+    send_message_to_users(
+        title=title,
+        content=content,
+        recipients=recipients,
+        permission_type=PermissionType.GROUP_MESSAGE,
+        image=image,
+        url=url,
+        tag=tag,
+        force_override_dont_send_email=True,
```

Comment:
Sending a message to a set of users creates that message for each individual user. When sending to a group we want to use the groups mailing list instead of sending to each individual email. If they choose to receive group messages as notifications we have to make when individual, but the email should still be to the list

Reply:
Hm, so if they are on the list, but choose to not get mail for group messages, but get push notification instead, that would work? This would not force an email to the users who have chosen push instead of mail?

Reply:
This would force emails to users as part of groups. We should set the settings for email for that notification type to `force_email = True` just to make it visible to the user.

The thought behind group messages is that they are the messages that we _have_ to send to groups no matter what. Messages that are sent to groups by onlineweb already, such as messages about new committee applications or 'this user was removed from an event after the deadline'-mails. Groups should not have a choice in getting those mails, as they are usually very important.

The group message part is basically just as the old mail system, just that individual users can choose to get those groups emails as push notifications as well.

Reply:
I see. Does this also mean e-mails sent to attendees for an event are forced through? Or are they not, since they are not a configured group?

Reply:
My thought were that the group option would be special, in sending to actual groups, like committees.

Event attendees are not groups, so this would not apply there. But that is a case where I think we would have to force an email anyways, since it is information that is important that is seen by the user.

Reply:
Personally, I like to have the option to be able to screw myself over, and not receiving possibly important information and rather take the consequence of it, if for example I feel that most of the event related e-mail is also sent to a Facebook-group or is really unnecessary. I am all for letting the user decide, and thus letting them also take the consequence, but that is a discussion for the actual event notification implementation, which I think is not part of this PR?

Reply:
Whatever we choose to do for personal emails that is something we can decide and change at runtime. At the very least we should put these things as defaults. When it comes to receipts we may have to send them for legal reasons, which means we do need the possibility to override a users settings.

I would also say that if we could avoid sending everything like that to our Facebook groups by using information channels like notifications it would be great in the end for us